api/admin.go

```diff
@@ -390,9 +405,79 @@ func (a *API) adminUserDelete(w http.ResponseWriter, r *http.Request) error {
 			return internalServerError("Error recording audit log entry").WithInternalError(terr)
 		}
 
-		if terr := tx.Destroy(user); terr != nil {
-			return internalServerError("Database error deleting user").WithInternalError(terr)
+		if params.IsSoftDelete == "true" {
+			if user.DeletedAt != nil {
+				// user has been soft deleted already
+				return nil
+			}
+			softDeleteId, terr := crypto.GenerateNanoId(5)
+			if terr != nil {
+				return terr
+			}
+			user.Email = storage.NullString(fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.GetEmail()))))
+			user.Phone = storage.NullString(fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.GetPhone()))))
+			user.EmailChange = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.EmailChange)))
+			user.PhoneChange = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(user.PhoneChange)))
+			now := time.Now()
+			user.DeletedAt = &now
+			if terr := tx.UpdateOnly(user, "email", "phone", "email_change", "phone_change", "deleted_at"); terr != nil {
+				return internalServerError("Error soft deleting user").WithInternalError(terr)
+			}
+
+			// set raw_user_meta_data to {}
+			userMetaDataUpdates := map[string]interface{}{}
+			for k := range user.UserMetaData {
+				userMetaDataUpdates[k] = nil
+			}
+			if terr := user.UpdateUserMetaData(tx, userMetaDataUpdates); terr != nil {
+				return internalServerError("Error soft deleting user meta data").WithInternalError(terr)
+			}
+
+			// set raw_app_meta_data to {}
+			appMetaDataUpdates := map[string]interface{}{}
+			for k := range user.AppMetaData {
+				appMetaDataUpdates[k] = nil
+			}
+			if terr := user.UpdateAppMetaData(tx, appMetaDataUpdates); terr != nil {
+				return internalServerError("Error soft deleting app meta data").WithInternalError(terr)
+			}
+
+			identities, terr := models.FindIdentitiesByUserID(tx, user.ID)
+			if terr != nil {
+				return internalServerError("Error retrieving identities").WithInternalError(terr)
+			}
+			// set identity_data to {}
+			for _, identity := range identities {
+				identity.ProviderId = fmt.Sprintf("%s-%x", softDeleteId, sha256.Sum256([]byte(identity.ProviderId)))
+				if terr := tx.UpdateOnly(identity, "provider_id"); terr != nil {
```

Comment:
I don't think there's a `provider_id` column on `identities` if looking at current master.

Reply:
it's here (https://github.com/supabase/gotrue/pull/489/files#diff-ea89c8e93d463b4f290c75a9e494c2dee36317457ae08cdf3282b0718f7a3acdR16)